Library/Homebrew/keg.rb

```diff
@@ -101,6 +102,19 @@ def initialize(path)
     @name = path.parent.basename.to_s
     @linked_keg_record = HOMEBREW_LIBRARY.join("LinkedKegs", name)
     @opt_record = HOMEBREW_PREFIX.join("opt", name)
+    begin
+      tab = Tab.from_file(path/"INSTALL_RECEIPT.json")
+      tap = tab.tap
+
+      oldname = if tap.nil? || tap == "Homebrew/homebrew"
+        Formulary.factory(name)
+      else
+        Formulary.factory("#{tap.sub("homebrew-", "")}/#{name}")
+      end.oldname
+
+      @oldname_opt_record = HOMEBREW_PREFIX.join("opt", oldname) if oldname
+    rescue Exception
+    end
```

Comment:
I mean can we get `oldname_opt_record` only using filesystem. 

>  so we have to exclude opt_record from opt_dir.subdirs before detecting.

Sorry I missed this part. How about:

```ruby
def oldname_opt_record
  @oldname_opt_record ||= if (opt_dir = HOMEBREW_PREFIX/"opt").directory?
    opt_dir.subdirs.detect { |dir| dir.symlink? && dir != opt_record && path == dir.resolved_path }
  end
end
```